composer/datasets/streaming/download.py

```diff
@@ -0,0 +1,132 @@
+    """Download a file from remote to local.
+
+    Args:
+        remote (str): Remote path (S3 or local filesystem).
+        local (str): Local path (local filesystem).
+        timeout (float): How long to wait for shard to download before raising an exception.
+    """
+    local_dir = os.path.dirname(local)
+    os.makedirs(local_dir, exist_ok=True)
+    start_time = time()
+
+    if remote.startswith('s3://'):
+        download_from_s3(remote, local, timeout=timeout)
+    else:
+        download_from_local(remote, local)
+
+    elapsed = time() - start_time
+    if elapsed > timeout:
+        raise TimeoutError(f'Waited too long (more than {timeout:.3f} sec) for download')
+
+
+def safe_download(remote: str, local: str, timeout: float = 20) -> None:
+    """Safely downloads a file from remote to local.
+       Handles multiple threads attempting to download the same shard.
+       Gracefully deletes stale tmp files from crashed runs.
+
+
+    Args:
+        remote (str): Remote path (S3 or local filesystem).
+        local (str): Local path (local filesystem).
+        timeout (float): How long to wait for shard to download before raising an exception. Default: 20 sec.
+    """
+    # If we already have the file cached locally, we are done.
+    if os.path.exists(local):
+        return
+
+    # Check if there is a tmp file.
+    local_tmp = local + '.tmp'
+    if os.path.exists(local_tmp):
+        local_tmp_create_time = os.path.getctime(local_tmp)
+        current_time = time()
+
+        if current_time - local_tmp_create_time < timeout + 1:  # 1s buffer to avoid race condition
+            # If the tmp file is recent, it is either (1) from a very recent crashed run, or (2) another thread is actively downloading it.
+            # So we wait but don't error out, in case we are in situation (1)
+            try:
+                wait_for_download(local, timeout)
+                return
+            except TimeoutError:
+                pass
+
+        # The tmp file is old, it is either (1) from a crashed run or (2) another thread is downloading it but is taking too long, and will timeout.
+        # Let's delete the tmp file. If situation (1), this is safe. If situation (2), the other thread is expected to crash with a TimeoutError anyways, so this is fine.
+        try:
+            os.remove(local_tmp)
+        except OSError:
+            # This occurs if another download thread got to the delete first.
+            pass
+
+    # There is no tmp file, so attempt to make it.
+    # If this fails, another download thread beat us to it, so wait.
+    # If we run out of time here, we know a download thread was active, so we should error out.
+    local_dir = os.path.dirname(local)
+    os.makedirs(local_dir, exist_ok=True)
+    try:
+        with open(local_tmp, 'xb') as out:
+            out.write(b'')
+    except FileExistsError:
+        wait_for_download(local, timeout)
```

Comment:
If this wait fails, should we automatically retry?

Reply:
Totally makes sense. Gonna add this to JIRA as it will require a refactor with multiprocessing locks, that are not currently implemented.